### DISCUSSION TITLE
Is LanguageDetector thread-safe?

### DISCUSSION THREAD

**Discussion by goldsam** [pemistahl/lingua-go#28] [Q&A]
Is `LanguageDetector` thread-safe?  I understand that different detectors share models behind the scene, but can `LanguageDetector` instances be shared between threads?

**Comment by pemistahl** [comment]
Hi Sam, I'm sorry for the late response. 

I'm quite sure it is thread-safe, even though I have not written the entire code with thread-safety in mind, to be honest. There is no global state within the library except for the data structures holding the language models in memory. The language models are stored in `sync.Map`, so they are safe to be used within multiple threads / goroutines. Each method call within the library creates its own set of variables to store the intermediate results of the algorithms. So it should be safe to use the same `LanguageDetector` instance within multiple threads / goroutines. It's definitely a good idea to write some tests to verify that. I put it on my todo list.
